Register and deregister services from ExoCom

hugobho opened the pull request:
As an ExoCom instance
  I want to be able to register and deregister services
  So that I can communicate with any new services that come online

@kevgo

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+
+  As an ExoCom instance
+  I want to be able to register and deregister services
+  So that I can communicate with any new services that come online
```
So that I can use all the services that are online at any given time

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+
+  Rules:
+  - ExoCom registers the service once it recieves an "exorelay.register" message from Exorelay
+  - ExoCom deregisters the service once it recieves an "exorelay.deregister" message from Exorelay
```
As I suggested in the PR on ExoRelay, this might go away

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+    When a new "users" service comes online at port 3001
+    Then it sends the following message to ExoCom:
+      """
+      * name: 'exorelay.register'
```
Let's call this `exocom.register-service`

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+          internal-namespace: 'foo'
+          host: 'localhost'
+          port: 3001
+          sends: ['message1']
```
I think a service shouldn't tell ExoCom which messages it sends and receives. It should just say "I am a user service, here is my certificate to prove it". ExoCom then knows what messages a service sends and receives, and subscribes the service instance to the right messages. This means ExoCom receives two types of messagse:

1. Information which services send and receive which messages. This is sent either upfront, or provided via command-line arguments

2. each service registers itself and what service type it is

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+        id: '123'
+      """
+    Then ExoCom now knows about these services:
+      | NAME           | INTERNAL NAMESPACE | HOST      | PORT |
```
mangled formatting

kevgo commented on features/steps/steps.ls:
```diff
@@ -77,6 +88,12 @@ module.exports = ->
     @set-service-landscape config, done
 
 
+  @When /^the "([^"]*)" service sends the following message to ExoCom:$/ (service, message-text, done) ->
```
... sends __this__ message to ExoCom:...

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,57 @@
+          receives: ['message2']
+        id: '123'
+      """
+    Then ExoCom no longer knows about the "users-service:localhost" service
```
Maybe its better to list all the services that ExoCom still knows about here

hugobho replied:
What do you mean by this?

trushton replied:
Ah okay, I was under the impression that there were other services still active. In that case, I personally think the way you did it makes the most sense.

hugobho commented:
Hey @kevgo can you take a look at this?

As of right now, services that come online do not send their messages configuration. Instead, ExoCom looks at a local `service-messages.yml` file which declares all the messages that a particular service receives. This is probably not the best way to configure which services send/receives which messages, but I figured that's a separate PR.

Also, services no longer send ZMQ messages to deregister themselves. A listener on their socket does that instead.

kevgo commented on src/client-registry/client-registry-spec.ls:
```diff
@@ -13,30 +13,34 @@ describe 'ClientRegistry', ->
+        service-name: 'tweets'
         internal-namespace: 'text-snippets'
-      result = @client-registry.external-message-name 'foo bar', service
+        }
```
2 spaces less indentation

kevgo commented on src/client-registry/client-registry.ls:
```diff
@@ -44,20 +45,39 @@ class ClientRegistry
+
+
+  _look-up-service-messages: (service-name) ->
+    service-messages = require ('../../service-messages.yml')
```
No parentheses necessary

kevgo commented:
Nice job Hugo! Once this wave of comments is addressed, this should be pretty much good to go.

hugobho commented:
Hey @kevgo, please take a look at the changes I made. I also addressed your comment about line 125 (now line 119) in the `steps.ls` file. Let me know what you think.

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,43 @@
+Feature: Manage new instances of services
+
+  As an Exosphere operator
+  I want that all service instances register themselves with ExoCom
```
This is grammatically not fully round. 

> I want that service instances can register themselves with ExoCom

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,43 @@
+  Scenario: a new service comes online
+    Given an ExoCom instance
+    When a new "users" service comes online at port 3001
+    Then it sends this message to ExoCom:
```
```cucumber
When a new service instance registers itself via the message:
  """
  ...
  """
Then ExoCom now knows about these service instances:

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,43 @@
+      """
+    And ExoCom now knows about these services:
+      | NAME           | TYPE  | INTERNAL NAMESPACE | HOST      | PORT |
+      | users-service  | users | foo                | localhost | 3001 |
```
Remove double space after `users-service`

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,43 @@
+
+
+  Scenario: deregister a service once it goes offline
+    Given an ExoCom instance
```
- use port 4xxx for testing (3xxx) is for running an app locally
- make sure that removing a service doesn't affect other service registrations

```cucumber
Given an ExoCom instance managing the service landscape:
  | NAME          | TYPE  | INTERNAL NAMESPACE | HOST      | PORT |
  | users-service | users | foo                | localhost | 4001 |
  | users-service | users | foo                | localhost | 4002 |
When the service instance at localhost:4001 goes offline
Then ExoCom now knows about these services:
  | NAME          | TYPE  | INTERNAL NAMESPACE | HOST      | PORT |
  | users-service | users | foo                | localhost | 4002 |

kevgo commented on features/support/mock-service.ls:
```diff
@@ -24,8 +24,10 @@ class MockService
+    | @closed => return
     @push-socket?.close!
     @pull-socket?.close!
+    @closed = true
```
Use `yes` instead of `true` here. It's the same value, but semantically more correct. "closed" is not truer than "open", but yes, it is closed now  :)

hugobho commented:
@kevgo

kevgo commented on features/register-service.feature:
```diff
@@ -0,0 +1,39 @@
+    When the "tweets-service" goes offline
+    Then ExoCom now knows about these services:
+      | NAME          | TYPE  | INTERNAL NAMESPACE | HOST      | PORT |
+      | users-service | users | foo                | localhost | 4001 |
```
This reads so well now!

kevgo commented on features/steps/steps.ls:
```diff
@@ -35,13 +36,32 @@ module.exports = ->
+      ..base-port = 5000
+      ..get-port N (@exocom-zmq-port) ~>
+      ..get-port N (@exocom-http-port) ~>
+        @create-exocom-instance http-port:@exocom-http-port, zmq-port: @exocom-zmq-port, ~>
```
add space after `:`

kevgo commented on features/steps/steps.ls:
```diff
@@ -35,13 +36,32 @@ module.exports = ->
+      ..get-port N (@exocom-zmq-port) ~>
+      ..get-port N (@exocom-http-port) ~>
+        @create-exocom-instance http-port:@exocom-http-port, zmq-port: @exocom-zmq-port, ~>
+          data or= []
```
`data` is not defined before, so is always `undefined` here.

kevgo commented on features/steps/steps.ls:
```diff
@@ -54,6 +74,17 @@ module.exports = ->
+
+
+  @When /^a new service instance registers itself via the message:$/ (config-text, done) ->
+    eval livescript.compile "config = \n#{config-text.replace /^/gm, '  '}", bare: yes, header: no
```
The name `config` is misleading for this variable. It contains a message sent to ExoCom. Call it `message`.

kevgo commented on src/client-registry/client-registry-spec.ls:
```diff
@@ -13,30 +13,34 @@ describe 'ClientRegistry', ->
     it "does not convert messages that don't match the format", ->
-      service =
-        name: 'tweets'
+      result = @client-registry.external-message-name {
```
Use `do` to get rid of the curly braces in this file?

kevgo commented on src/message-sender/message-sender.ls:
```diff
@@ -7,16 +7,34 @@ debug = require('debug')('exocom:message-sender')
     @service-sockets = {}
 
+    # Caches message metadata (id, timestamp) to calculate response time
+    @message-cache = new MessageCache!
```
Where does this line come from? Do you need to sync this branch?
